AlgebraInLean/World0LeanIntroSOLUTIONS/Sheet2Solutions.lean

```diff
@@ -0,0 +1,283 @@
+  exact h1
+  done
+
+/-NNG Implication World Level 2-/
+/-hint: zero_add still works outside of the NNG, as it is a theorem proven in mathlib.-/
+example (y x : Nat) (h1 : 0 + x = 0 + y + 2) : x = y + 2 := by
+  rw [zero_add] at h1
+  rw [zero_add] at h1
+  exact h1
+  done
+
+/-NNG Implication World Level 3-/
+example (x y : Nat) (h1 : x = 37) (h2 : x = 37 → y = 42) : y = 42 := by
+  apply h2 at h1
+  exact h1
+  done
+
+/-NNG Implication World Level 6-/
+example (x : Nat) : x = 37 → x = 37 := by
+  intro h1
+  exact h1
+  done
+
+/-Let's move on beyond the NNG to do some exercises you haven't seen before. The following exercises
+are from Kevin Buzzard's Formalizing Mathematics.-/
+
+variable (P Q R S T : Prop)
+/-Note that the variables we are working with here are of the type Prop.-/
+
+/-FM Section 1 Sheet 1-/
+example : P → P := by
+  intro h
+  exact h
+  done
+
+/-FM Section 1 Sheet 1-/
+example : P → Q → P := by
+  intro h1
+  intro h2
+  exact h1
+  done
+
+/-FM Section 1 Sheet 1-/
+example : P → (P → Q) → Q := by
+  intro h1 h2
+  apply h2 at h1
+  exact h1
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → Q) → (Q → R) → P → R := by
+  intro h1 h2 h3
+  apply h1 at h3
+  apply h2 at h3
+  exact h3
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → Q → R) → (P → Q) → P → R := by
+  intro h1 h2 h3
+  apply h1
+  exact h3
+  apply h2 at h3
+  exact h3
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → R) → (S → Q) → (R → T) → (Q → R) → S → T := by
+  intro h1 h2 h3 h4 h5
+  apply h2 at h5
+  apply h4 at h5
+  apply h3 at h5
+  exact h5
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → Q) → ((P → Q) → P) → Q := by
+  intro h1 h2
+  apply h1
+  apply h2
+  exact h1
+  done
+
+/-FM Section 1 Sheet 1-/
+example : ((P → Q) → R) → ((Q → R) → P) → ((R → P) → Q) → P := by
+  intro h1 h2 h3
+  apply h2
+  intro h4
+  apply h1
+  intro h5
+  exact h4
+  done
+
+/-FM Section 1 Sheet 1-/
+example : ((Q → P) → P) → (Q → R) → (R → P) → P := by
+  intro h1 h2 h3
+  apply h1
+  intro h4
+  apply h3
+  apply h2
+  exact h4
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (((P → Q) → Q) → Q) → P → Q := by
+  intro h1 h2
+  apply h1
+  intro h3
+  apply h3
+  exact h2
+  done
+
+/-FM Section 1 Sheet 1-/
+example :
+    (((P → Q → Q) → (P → Q) → Q) → R) →
+      ((((P → P) → Q) → P → P → Q) → R) → (((P → P → Q) → (P → P) → Q) → R) → R := by
+  intro h1 h2 h3
+  apply h2
+  intro h4 h5 h6
+  apply h4
+  intro h7
+  exact h5
+  done
+
+/-The ↔ means "if and only if." If you are ever curious how to type a certain symbol in Lean, just
+hover over it for a few seconds. rw can be applied to ↔ hypotheses.-/
+
+/-FM Section 1 Sheet 5-/
+example : (P ↔ Q) → (Q ↔ P) := by
+  intro h
+  rw [h]
+  done
+
+/-FM Section 1 Sheet 5-/
+example : (P ↔ Q) → (Q ↔ R) → (P ↔ R) := by
+  intro h1 h2
+  rw [h2] at h1
+  exact h1
+  done
+
+/-Note that many of the exercises above could be solved in one line, using the tactic "tauto," just
```

Comment:
It'd be fun to explain where "tauto" comes from, so you don't have idiots like me pronouncing it T-auto!